### DISCUSSION TITLE
Should flash-loanable funds be opt-in?

### DISCUSSION THREAD

**Discussion by razgraf** [sablier-labs/evm-monorepo#97] [General]
For streams with tokens that are very illiquid you might be able to use flash-loans with Sablier (with funds coming from vesting streams) to short the asset or arbitrage it pretty easily.

Should we integrate or develop an opt-in system?
Should this system be per-stream or per-asset?

**Comment by PaulRBerg** [comment]
> to short the asset

You can't actually short anything in practice because the trade would be undone by the time the flash loan is terminated.

> arbitrage it pretty easily

This could indeed be a problem in the sense that token holders wouldn't expect there to be arbitrageurs that source liquidity from vesting contracts.

> Should we integrate or develop an opt-in system?

My take on this is the following:

The real reason not to allow global flash loaning is simply this: a vesting contract is supposed to be fully locked. Unlocking the capital, even for just 1 block, would break that assumption. To make an analogy, think about how absurd it would be to arbitrage some equity in a company, which is being vested. You would sell it on exchange A (NASDAQ) and buy it back on exchange B (NYSE). Makes no sense.

Thus, I think that by default **all tokens** should be non-flash-loanable.

**Comment by razgraf** [comment]
> You can't actually short

In theory, shorting is a part of the arbitrage, so the first two points you commented on are still both valid (e.g. borrow asset, sell asset, perform exploit, drop asset price, buy asset, repay, pocket the difference). Be it a single-block pressure on the price, but still a short.

> The real reason not to allow global flash loaning is simply this: a vesting contract is supposed to be fully locked

On your flash loans vs. vesting comment, I'd argue flash loans are themselves a new concept so it's hard to imagine them in the real economy. On the other hand, I do share your opinion on the assumption the user makes when fully locking tokens in a stream - with a primary reason being getting them out of circulation.

All in all, I still believe flash loans would (1) become a good source of fees for the protocol (2) add an extra differentiating feature, as a liquidity source for arbitrageurs. For the latter, loaning $USDC proves economically safer than the illiquid $TOKEN so this is where we might need some finer control over which assets are enabled on not.

**Comment by PaulRBerg** [comment]
> so the first two points you commented on are still both valid (e.g. borrow asset, sell asset, perform exploit, drop asset price, buy asset, repay, pocket the difference).

This assumes that there is a lending market set up for the vested token, which is not always the case. But I generally agree with you.

> with a primary reason being getting them out of circulation.

Yup.

> All in all, I still believe flash loans would (1) become a good source of fees for the protocol (2) add an extra differentiating feature, as a liquidity source for arbitrageurs.

Agree.

> For the latter, loaning $USDC proves economically safer than the illiquid $TOKEN so this is where we might need some finer control over which assets are enabled on not.

We should internally give the tokens some "measure" of flash-loan-ability. Stablecoins would be given 1 (maximum possible) and newly printed tokens would be given 0 (lowest possible).

**Comment by PaulRBerg** [comment]
We have agreed upon not making all tokens flash-loanable by default (it will be a per-token mapping that controls this).

Locking this conversation in favor of https://github.com/sablierhq/v2-core/discussions/170.
